File: Chapter_5/src/main/scala/udf.scala
// import packages

import org.apache.spark.sql.SparkSession

// defining a singleton scala object containing the main method

object udf{
  def main(args:Array[String]){

    // creating a spark session

    val spark = SparkSession.builder.appName("udf_creation").getOrCreate()

    // defining a UDF

    val cube = (a : Long) => { a * a * a }

    // register UDF

    spark.udf.register("cubed",cube)

    // creating a temporary view with random data
    //
    // creating a dataframe using sequence

    val data =  Seq((1,"john","smith"),(2,"tell","tale"))

    val df = spark.createDataFrame(data).toDF("id","first_name","last_name")

    df.show()

    spark.range(1,9).createOrReplaceTempView("udf_test")

    spark.sql("""SELECT id, cubed(id) AS id_cubed FROM udf_test""").show()









    // stop spark session
    spark.stop()
  }
}

File: Chapter_5/pyspark/udf.py
# import packages

from pyspark.sql import SparkSession
from pyspark.sql.functions import *

# creating an entry point of the application
if __name__ == "__main__":

    # defining a UDF
    def cube (a):
        return a*a*a
    
    # creating an instance of sparksession
    spark = SparkSession.builder.appName("udf_examples").getOrCreate()

    # register the UDF
    spark.udf.register("cube_it",cube)

    # defining a sample dataframe
    data_values = [[1,"john","smith"], [2,"adam","apple"], [3,"george","washington"]]
    
    df = spark.createDataFrame(data_values, ("id","first_name","last_name"))

    # creating a view from it
    df.createOrReplaceTempView("view_for_udf")

    # performing sparksql query including a UDF operation
    spark.sql("""SELECT id, cube_it(id) AS cubed_id FROM view_for_udf""").show()

    #stop the spark session
    spark.stop()

